File: days/src/main/java/aoc2024j/day2/Day2.java

package aoc2024j.day2;

import utils.IO;

import java.util.AbstractList;
import java.util.List;
import java.util.stream.Gatherers;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class Day2 {

    record Interval(int min, int max) {
        boolean contains(int value) {
            return min <= value && value <= max;
        }
    }

    // step is defined as x1 - x0
    static Interval INCREASING = new Interval(1, 3);
    static Interval DECREASING = new Interval(-3, -1);

    record Report(List<Integer> levels) {
        static Report parse(String line) {
            var levels = Stream.of(line.split(" "))
                    .map(Integer::parseInt)
                    .toList();
            return new Report(levels);
        }

        boolean isValid() {
            return containsAllDifferences(levels, INCREASING)
                    || containsAllDifferences(levels, DECREASING);
        }

        private boolean containsAllDifferences(List<Integer> levels, Interval interval) {
            return levels.stream()
                    .gather(Gatherers.windowSliding(2))
                    .allMatch(pair -> interval.contains(pair.get(1) - pair.get(0)));
        }

        boolean isValidWithHoles() {
            return isValid() || IntStream.range(0, levels.size())
                    .mapToObj(i -> new Report(new HoledList<>(levels, i)))
                    .anyMatch(Report::isValid);
        }
    }

    long part1(List<String> data) {
        return data.stream()
                .map(Report::parse)
                .filter(Report::isValid)
                .count();
    }

    static class HoledList<E> extends AbstractList<E> {
        private final List<E> list;
        private final int hole;

        HoledList(List<E> list, int hole) {
            assert 0 <= hole && hole < list.size();
            this.list = list;
            this.hole = hole;
        }

        @Override
        public E get(int index) {
            return list.get(index < hole ? index : index + 1);
        }

        @Override
        public int size() {
            return list.size() - 1;
        }
    }

    long part2(List<String> data) {
        return data.stream()
                .map(Report::parse)
                .filter(Report::isValidWithHoles)
                .count();
    }

    public static void main(String[] args) {
        var day2 = new Day2();
        var data = IO.getResourceAsList("aoc2024/day2.txt");
        var part1 = day2.part1(data);
        System.out.println("part1 = " + part1);
        var part2 = day2.part2(data);
        System.out.println("part2 = " + part2);
    }
}

File: days/src/main/scala/aoc2024s/day2/Day2.scala

package aoc2024s.day2;

import utils.IO;
import scala.jdk.CollectionConverters.*

object Day2 {

  case class Interval(min: Int, max: Int) {
    def contains(value: Int): Boolean = min <= value && value <= max
  }

  val INCREASING: Interval = Interval(1, 3)
  val DECREASING: Interval = Interval(-3, -1)

  case class Report(levels: List[Int]) {
    def isValid: Boolean =
      containsAllDifferences(levels, INCREASING)
        || containsAllDifferences(levels, DECREASING)

    private def containsAllDifferences(levels: List[Int], interval: Interval): Boolean =
      levels.sliding(2).forall {
        case List(a, b) => interval.contains(b - a)
      }

    private def makeHole(pos: Int): Report =
      Report(levels.take(pos) ++ levels.drop(pos + 1))

    def isValidWithHoles: Boolean =
      isValid || levels.indices.exists(makeHole(_).isValid)
  }

  object Report {
    def parse(line: String): Report = {
      val levels = line.split(" ").map(_.toInt).toList
      Report(levels)
    }
  }

  def part1(data: List[String]): Long =
    data.map(Report.parse).count(_.isValid)

  def part2(data: List[String]): Long = {
    data.map(Report.parse).count(_.isValidWithHoles)
  }

  @main def main2(): Unit = {
    val data = IO.getResourceAsList("aoc2024/day2.txt").asScala.toList;
    val part1 = Day2.part1(data);
    println(s"part1 = $part1");
    val part2 = Day2.part2(data);
    println(s"part2 = $part2");
  }
}